platform/consul/conversion.go

```diff
@@ -0,0 +1,149 @@
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/api"
+	"istio.io/pilot/model"
+)
+
+const (
+	protocolTagName = "protocol"
+	externalTagName = "external"
+)
+
+func convertTags(tags []string) model.Tags {
+	out := make(model.Tags, len(tags))
+	for _, tag := range tags {
+		vals := strings.Split(tag, "=")
+
+		// Tags not of form "key=value" are ignored to avoid possible collisions
+		if len(vals) > 1 {
+			out[vals[0]] = vals[1]
+		}
+	}
+	return out
+}
+
+func convertPort(port int, name string) *model.Port {
+	if name == "" {
+		name = "http"
+	}
+
+	return &model.Port{
+		Name:     name,
+		Port:     port,
+		Protocol: convertProtocol(name),
+	}
+}
+
+func convertService(endpoints []*api.CatalogService) *model.Service {
+	name, addr, external := "", "", ""
+	node, datacenter := "", ""
+
+	ports := model.PortList{}
+	for _, endpoint := range endpoints {
+		name = endpoint.ServiceName
+		ports = append(ports, convertPort(endpoint.ServicePort, endpoint.NodeMeta[protocolTagName]))
+
+		// TODO This will not work if service is a mix of external and local services
+		// or if a service has more than one external name
+		if endpoint.NodeMeta[externalTagName] != "" {
+			external = endpoint.NodeMeta[externalTagName]
+		}
+
+		// TODO how to handle if a service spans datacenters/nodes?
+		node = endpoint.Node
+		datacenter = endpoint.Datacenter
+
+	}
+
+	out := &model.Service{
+		Hostname:     serviceHostname(name, node, datacenter),
+		Ports:        ports,
+		Address:      addr,
+		ExternalName: external,
+	}
+
+	return out
+}
+
+func convertInstance(inst *api.CatalogService) *model.ServiceInstance {
+	tags := convertTags(inst.ServiceTags)
+	port := convertPort(inst.ServicePort, inst.NodeMeta[protocolTagName])
+
+	addr := inst.ServiceAddress
+	if addr == "" {
+		addr = inst.Address
+	}
+
+	return &model.ServiceInstance{
+		Endpoint: model.NetworkEndpoint{
+			Address:     addr,
+			Port:        inst.ServicePort,
+			ServicePort: port,
+		},
+
+		Service: &model.Service{
+			Hostname: serviceHostname(inst.ServiceName, "TODO", inst.Datacenter),
+			Address:  inst.ServiceAddress,
+			Ports:    model.PortList{port},
+			// TODO ExternalName come from metadata?
+			ExternalName: inst.NodeMeta[externalTagName],
+		},
+		Tags: tags,
+	}
+}
+
+// serviceHostname produces FQDN for a consul service
+func serviceHostname(name, node, datacenter string) string {
+	// TODO include consul node in Hostname?
+	// consul DNS uses "redis.service.us-east-1.consul" -> "[<optional_tag>].<svc>.service.[<optional_datacenter>].consul"
+	return fmt.Sprintf("%s.service.consul", name)
+}
+
+// parseHostname extracts service name from the service hostname
+func parseHostname(hostname string) (name string, err error) {
+	parts := strings.Split(hostname, ".")
+	if len(parts) < 1 {
+		err = fmt.Errorf("missing service name from the service hostname %q", hostname)
+		return
+	}
+	name = parts[0]
+	return
+}
+
+func convertProtocol(name string) model.Protocol {
```

Comment:
@kyessenov The k8s convert function is `convertProtocol(name string, proto v1.Protocol)` which does conversion based on protocol and port name.  Consul doesn't have a concept of protocol in its service instance definition or a port name so I am not sure it maps well to use the same function